rclpy/rclpy/executors.py

```diff
@@ -0,0 +1,290 @@
+            client.client_handle, client.srv_type.Response, client.sequence_number)
+        return response
+
+    def _execute_client(self, client, response):
+        if response:
+            # clients spawn their own thread to wait for a response in the
+            # wait_for_future function. Users can either use this mechanism or monitor
+            # the content of client.response to check if a response has been received
+            client.response = response
+
+    def _take_service(self, srv):
+        request_and_header = _rclpy.rclpy_take_request(
+            srv.service_handle, srv.srv_type.Request)
+        return request_and_header
+
+    def _execute_service(self, srv, request_and_header):
+        if request_and_header is None:
+            return
+        (request, header) = request_and_header
+        if request:
+            response = srv.callback(request, srv.srv_type.Response())
+            srv.send_response(response, header)
+
+    def _make_handler(self, entity, take_from_wait_list, call_callback):
+        """Make a handler that performs work on an entity."""
+        gc = self._guard_condition
+        # Mark this so it doesn't get added back to the wait list
+        entity._executor_event = True
+
+        def handler():
+            nonlocal entity
+            nonlocal gc
+            if not entity.callback_group.take(entity):
+                # Didn't get the callback, try again later
+                entity._executor_event = False
+                _rclpy.rclpy_trigger_guard_condition(gc)
+                return
+            try:
+                arg = take_from_wait_list(entity)
+
+                # Signal that this has been 'taken' and can be added back to the wait list
+                entity._executor_event = False
+                _rclpy.rclpy_trigger_guard_condition(gc)
+
+                call_callback(entity, arg)
+            finally:
+                entity.callback_group.give_back(entity)
+                # Signal that work has been done so the next callback in a mutually exclusive
+                # callback group can get executed
+                _rclpy.rclpy_trigger_guard_condition(gc)
+        return handler
+
+    def _filter_eligible_entities(self, entity_list):
+        """Filter entities that should not be put onto the wait list."""
+        return [e for e in entity_list if e.callback_group.can_take(e) and not e._executor_event]
+
+    def wait_for_ready_callbacks(self, timeout_sec=None, nodes=None):
+        """
+        Yield callbacks that are ready to be performed.
+
+        timeout_sec - seconds to wait for a callback. Blocks forever if None. Don't wait if <= 0
+        nodes - a subset of the added nodes to wait on. Wait on all nodes if None.
+
+        It's possible for this to return before the timeout without yielding any callbacks. In this
+        case the executor should call this method again.
+
+        Yeilds 3-tuple (callable, entity, node)
```

Comment:
Spelling: Yields

Reply:
Fixed in 7b4684c17d211a387a79ebd64608906b4e352f91